### PR TITLE
[bugfix] Fixes regression in relation reference widget with 'chain filter' option

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

 This PR fixes a regression in the relation reference widget when the `chain filters` option is activated.

 See https://github.com/qwat/QWAT/issues/190#issuecomment-380454685 for a GIF highlighting the issue.

 Some tests have been added.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/gui/editorwidgets/qgsrelationreferencewidget.h [hunk 1]
```diff
@@ -183,6 +183,7 @@ class GUI_EXPORT QgsRelationReferenceWidget : public QWidget
     void filterChanged();
     void addEntry();
     void updateAddEntryButton();
+    void updateIndex();
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.h]: Can you add a short description to this method (even though it's not in the API docs)? It's not trivial to understand what it does (at least for me).

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp [hunk 2]
```diff
@@ -131,7 +131,11 @@ void QgsValueRelationWidgetWrapper::initWidget( QWidget *editor )
     mTableWidget->setEditTriggers( QAbstractItemView::NoEditTriggers );
     mTableWidget->setSelectionMode( QAbstractItemView::NoSelection );
     if ( mCache.size() > 0 )
-      mTableWidget->setRowCount( ( mCache.size() + config( QStringLiteral( "NofColumns" ) ).toInt() - 1 ) / config( QStringLiteral( "NofColumns" ) ).toInt() );
+    {
+      const int nofCols = config( QStringLiteral( "NofColumns" ) ).toInt();
+      const int denom = nofCols != 0 ? nofCols : 1;
+      mTableWidget->setRowCount( ( mCache.size() + nofCols - 1 ) / denom );
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Is this related?

File: src/gui/editorwidgets/qgsrelationreferencewidget.h [hunk 3]
```diff
@@ -213,6 +214,7 @@ class GUI_EXPORT QgsRelationReferenceWidget : public QWidget
     bool mShown = false;
     QgsRelation mRelation;
     bool mIsEditable = true;
+    QVariant mNullIndex;
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.h]: This is to define if the null index is valid? I find the type/variable name combination hard to interpret. The name looks like it's an integer, the type is flexible and I think in reality it's a `std::optional<bool>` which we cannot use (yet). What do you think about using a private enum (or at least a name change) here?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp (hunk 2 above)
Comment (pblottiere) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: No it's not related but I came across this issue with my testing project. If you prefer, I can create a dedicated PR?

File: src/gui/editorwidgets/qgsrelationreferencewidget.h (hunk 3 above)
Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: > I think in reality it's a std::optional<bool> which we cannot use (yet)

Exactly. 

> What do you think about using a private enum (or at least a name change) here?

I think that a private enum would be overkill in this case, but i totally can change the name. What do you think about `mUpdateToNullIndex`. But maybe do you have a better idea in mind?

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: I'm not even completely sure what it does. Would `mNullIndexRequiredUpdate` be more precise?

Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: > I'm not even completely sure what it doe

The aim is to have the same behavior than QGIS 2.18 when a user is changing the filter by clicking on comboboxes. For example:

untitled

- if the user selects `diameter`, then the FK is set to the first valid result (or `NULL` if null values are allowed). Moreover, the `material_value` combobox is not available anymore while a value is not selected in `diameter` (but it's not relevant in this case).
- if the user selects a value in `material_value` combobox, then the corresponding filter is applied and FK is the first valid value

Thus, the current `mNullIndex` allow to indicate if we're in the 1st case or the second case.

By hoping I'm not too unclear........

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp (hunk 2 above)
Comment (m-kuhn) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Ah no, that looks ok.

File: src/gui/editorwidgets/qgsrelationreferencewidget.h (hunk 3 above)
Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: Sorry, I'm completely on the wire. Can you help me with a small list explaining when `mNullIndex` is set to true, to false and to Null? It looks like it's only set to Null at initialization and after `modelUpdated` (where it has been evaluated and used), so the Null state of the mNullIndex is probably a guard against multiple runs of this codeblock?

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 4]
```diff
@@ -844,6 +873,21 @@ void QgsRelationReferenceWidget::filterChanged()
     }
   }
   filterExpression = filters.values().join( QStringLiteral( " AND " ) );
+
+  if ( mChainFilters && mComboBox->count() > 0 )
+  {
+    mNullIndex.clear();
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Is this required? The value will be set anyway either in `if` or in `else`, no?

Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Indeed, you're right.

File: src/gui/editorwidgets/qgsrelationreferencewidget.h (hunk 3 above)
Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: Yes, the Null state is a guard. Actually, as soon as the underlying model is updated,  the new slot `updateIndex` is called. However, I don't want to modify the current FK each time, but just when the model has been updated due to a user action on comboboxes. 

This way:
- `mNullIndex` is always Null except when the user click on comboboxes. So I know that I have to update the FK once the model is updated
- `mNullIndex` is True when the user click on a first item in one of the combobox (so on `brides`, `diameter` or `material_value` in the above example). So I know that I have to set the FK to Null or the first valid value (according to the mAllowValue configuration).
- `mNullIndex` is False when the user click on a value in one of the combobox. So I know that I have to select the first valid FK.

Retrospectively, I think that the member `mNullIndex` is really badly named and very confusing...

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: Ah, I see. I think the variable name should somehow include the term `Filter` or `ChainFilter` and possibly also a descriptive comment (and a reminder to migrate to `std::optional`, once available ;) )

I didn't test it, just according to the description above, is the following interpretation of a possible workflow correct?

1. Select the bride 120, it will be set to the first valid entry (let's assume it's 9)
2. Select the material_value 'PVC' it's set to a possibly different entry (let's assume 10)
3. Unselect material_value again, it's set to NULL (even though the filter state is the same as in step 1)

Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: >  I think the variable name should somehow include the term Filter or ChainFilter and possibly also a descriptive comment (and a reminder to migrate to std::optional, once available ;) )

OK, I'll do that :).

> Unselect material_value again, it's set to NULL (even though the filter state is the same as in step 1)

Yes, it's the current behavior on 2.18 branch.

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: And is that desired behavior?

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: As the original author of this behavior, I can say that this was more a side-effect than intended behavior @pblottiere , if you have a chance to adjust that I don't think it will break anyones workflow but instead make the widget more predictable.

Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: > if you have a chance to adjust that I don't think it will break anyones workflow but instead make the widget more predictable.

OK, I'll do that.

Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.h]: Hi @m-kuhn! 

I updated the code accordingly. This way, the `mNullIndex` guard is not necessary anymore :).

Let me know.

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 5]
```diff
@@ -155,6 +156,38 @@ QgsRelationReferenceWidget::~QgsRelationReferenceWidget()
...
 }
 
+void QgsRelationReferenceWidget::updateIndex()
+{
+  if ( mChainFilters && mComboBox->count() > 0 )
+  {
+    int index = -1;
+
+    // uninitialized filter
+    if ( mFilterComboBoxes.count() > 0
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: `mFilterComboBoxes` is a `QList`? Use `isEmpty()` instead (it's more performant).

File: src/gui/qgsfeaturelistcombobox.cpp [hunk 6]
```diff
@@ -136,6 +137,18 @@ void QgsFeatureListComboBox::restoreLineEditState()
     mLineEditState.restore( mLineEdit );
 }
 
+int QgsFeatureListComboBox::nullIndex() const
+{
+  int index = -1;
+
+  if ( allowNull() )
+  {
+    index = findText( tr( "NULL" ) );
```
Comment (m-kuhn) [inline src/gui/qgsfeaturelistcombobox.cpp]: Is this reliable if there's an item with the text `NULL` but which is *not* the NULL value?

**Comment by m-kuhn** [comment]
Thanks, looks much simpler, I like :+1:

File: src/gui/qgsfeaturelistcombobox.cpp (hunk 6 above)
Comment (pblottiere) [reply src/gui/qgsfeaturelistcombobox.cpp]: Considering that `findText` will search for the first element matching NULL text and that the "true" NULL value is always added ahead of others items, I think so.

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp (hunk 5 above)
Comment (pblottiere) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Of course, silly me...

Comment (nyalldawson) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: #pedantic

I don't think it's actually faster (I looked into this once), but it *is* much easier to understand the logic behind the condition by using empty() instead.

Other dumb #pedantic thing: if you use empty() instead of isEmpty(), it'll be easier for us to change to std containers in future...

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Totally unrelated comment:

In this case it's `QgsVectorLayer::empty()` and not `QgsVectorLayer::isEmpty()` which I'm going to add soon.

Comment (nyalldawson) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: QgsFeatureSource::empty()? (pretty please :angel: )

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Hehe, I'll have a look but can't guarantee because I'm not sure it's trivial to know for all feature sources.

Either

 * Introduce `QgsFeatureSource::capabilities()` (`KnowsIfItIsEmpty`)
 * Introduce `QgsFeatureSource::canDetermineEmpty()`
 * The default implementation consumes the whole source until it either reaches the end or finds an entry. Potentially expensive and might leave resources hanging.
 * If in doubt define if `false` or `true` should be returned
 * Use the unavailable `std::optional<bool>`

Comment (nyalldawson) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: There's already QgsFeatureSource::featureCount - that could be used as a fallback?

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: If that works, `empty()` works as well :+1:

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: > There's already QgsFeatureSource::featureCount - that could be used as a fallback?

Looking at this in detail, the question remains open. What to do when featureCount() returns -1

Comment (nyalldawson) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Return an enum? Empty/NotEmpty/Unknown?

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: I'll intuitively always write `if ( layer->empty() ) ` because I'm used to a method called `empty()` to return a boolean.
And depending on the feature source, this is a safe thing to do (e.g. if I know it's a pg layer).

Maybe NotEmpty = 0, Empty = 1, Unknown = 2 would work.

Comment (nyalldawson) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: I don't like that, because it means unknown state layers will be skipped by a `if ( layer->empty() )` check. Maybe we should just treat -1 as not empty, since that's going to be the far more common case?

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: In this case we'll at least want a different method that let's us query the unknown state which the dev is free (not to) use.
